src/runtime.rs

```diff
@@ -0,0 +1,371 @@
+/// You can obtain a handle to spawn tasks on the runtime, which is cheaply cloneable.
+///
+/// The runtime has a shutdown method that will wait for some time all tasks to finish.
+pub mod tpc {
+    use futures::{future::LocalBoxFuture, Future, FutureExt};
+    use std::{
+        fmt,
+        sync::{Arc, Mutex},
+        time::Duration,
+    };
+
+    /// A wrapper to manage multiple tokio runtimes in a thread per core fashion.
+    pub struct Runtime {
+        name: String,
+        /// The handle to spawn tasks on the runtime
+        handle: Handle,
+        /// The handles to the threads
+        handles: Vec<std::thread::JoinHandle<()>>,
+    }
+
+    impl fmt::Debug for Runtime {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            f.debug_struct("Runtime")
+                .field("name", &self.name)
+                .field("threads", &self.handles.len())
+                .finish_non_exhaustive()
+        }
+    }
+
+    /// The handle to spawn tasks on the runtime
+    #[derive(Clone)]
+    pub struct Handle {
+        sender: Arc<Mutex<Option<flume::Sender<Task>>>>,
+    }
+
+    impl fmt::Debug for Handle {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            f.debug_struct("Handle").finish()
+        }
+    }
+
+    struct Task(Box<dyn FnOnce() -> LocalBoxFuture<'static, ()> + Send + 'static>);
+
+    impl Runtime {
+        pub fn new(name: &str, n: usize) -> Self {
+            let name = name.to_string();
+            let (task_sender, task_receiver) = flume::bounded::<Task>(1);
+            let handles = (0..n)
+                .map(|i| {
+                    let task_receiver = task_receiver.clone();
+                    // name for the non blocking thread that we spawn
+                    let main_name = format!("{}-{}", name, i);
+                    // name for the blocking thread that the runtime spawns
+                    let blocking_name = format!("{}-{}-blocking", name, i);
+                    // spawn the thread
+                    std::thread::Builder::new()
+                        .name(main_name.clone())
+                        .spawn(move || {
+                            let local = tokio::task::LocalSet::new();
+                            let rt = tokio::runtime::Builder::new_current_thread()
+                                .max_blocking_threads(1)
+                                .thread_name(blocking_name)
+                                .build()
+                                .expect("failed to build tokio runtime");
+                            rt.block_on(local.run_until(async move {
+                                while let Ok(Task(f)) = task_receiver.recv_async().await {
+                                    tokio::task::spawn_local(f());
+                                }
+                            }));
+                            tracing::trace!("runtime {} dropped", main_name);
```

Comment:
this should probably do some `span` tracing magic to make debugging easier, cc @flub

Reply:
Not sure you want to create a span for the entire duration where the runtime lives. It wouldn't add much info.

I can remove this trace, since I just used it in debugging of tests.

Reply:
Sorry, I meant around the stuff in the `spawn_local`, so it is clear which runtime is executing sth

Reply:
Ah, I see. I can do this. But is it worth it? I would have to clone the name for each task. And what does it gain you? If you want to know on which single threaded runtime you are running you can always output the thread name.

Reply:
true, lets leave it as is for now